jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/compress/ByteAccumulator.java

```diff
@@ -19,56 +19,92 @@
 
     public ByteAccumulator(int maxOverallBufferSize)
+    {
+        this(maxOverallBufferSize, null);
+    }
+
+    public ByteAccumulator(int maxOverallBufferSize, ByteBufferPool byteBufferPool)
     {
         this.maxSize = maxOverallBufferSize;
+        this.accumulator = new ByteBufferAccumulator(byteBufferPool, false);
     }
 
-    public void copyChunk(byte[] buf, int offset, int length)
+    public int getLength()
     {
-        if (this.length + length > maxSize)
+        return length;
+    }
+
+    public ByteBuffer ensureBuffer(int minAllocationSize)
+    {
+        return accumulator.ensureBuffer(MIN_SPACE, minAllocationSize);
+    }
+
+    public void addLength(int read)
+    {
+        length += read;
+        if (length > maxSize)
         {
-            String err = String.format("Resulting message size [%,d] is too large for configured max of [%,d]", this.length + length, maxSize);
+            String err = String.format("Resulting message size [%d] is too large for configured max of [%d]", length, maxSize);
             throw new MessageTooLargeException(err);
         }
-
-        byte[] copy = new byte[length - offset];
-        System.arraycopy(buf, offset, copy, 0, length);
-
-        chunks.add(copy);
-        this.length += length;
     }
 
-    public int getLength()
+    public void copyChunk(byte[] buf, int offset, int length)
     {
-        return length;
+        copyChunk(BufferUtil.toBuffer(buf, offset, length));
     }
 
-    public void transferTo(ByteBuffer buffer)
+    public void copyChunk(ByteBuffer buffer)
     {
-        if (buffer.remaining() < length)
+        int remaining = buffer.remaining();
+        if (length + remaining > maxSize)
         {
-            throw new IllegalArgumentException(String.format("Not enough space in ByteBuffer remaining [%d] for accumulated buffers length [%d]",
-                buffer.remaining(), length));
+            String err = String.format("Resulting message size [%d] is too large for configured max of [%d]", length + remaining, maxSize);
+            throw new MessageTooLargeException(err);
         }
 
-        int position = buffer.position();
-        for (byte[] chunk : chunks)
+        length += remaining;
+        accumulator.copyBuffer(buffer);
+    }
+
+    public void transferTo(ByteBuffer buffer)
+    {
+        // For some reason this method expects the buffer in fill mode but returns a buffer in flush mode.
+        BufferUtil.flipToFlush(buffer, 0);
+
+        int availableSpace = BufferUtil.space(buffer);
+        if (availableSpace < length)
         {
-            buffer.put(chunk, 0, chunk.length);
+            String err = String.format("Not enough space in ByteBuffer remaining [%d] for accumulated buffers length [%d]", availableSpace, length);
+            throw new IllegalArgumentException(err);
         }
-        BufferUtil.flipToFlush(buffer, position);
+
+        accumulator.writeTo(buffer);
+        close();
```

Comment:
`close()` method invoked by [DeflateFrameExtension.java#L64](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x-ByteAccumulator/jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/compress/DeflateFrameExtension.java#L64)
and [PerMessageDeflateExtension.java#L81](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x-ByteAccumulator/jetty-websocket/websocket-common/src/main/java/org/eclipse/jetty/websocket/common/extensions/compress/PerMessageDeflateExtension.java#L81)
`transferTo` method no need add extra `close` method

Reply:
I did this to preserve the behaviour of this class with the previous usage.
Before these changes it was not AutoCloseable, so it would have been used like this with no closing:
```java
ByteAccumulator acc = new ByteAccumulator();
acc.copyChunk(something);
...
acc.transferTo(combinedBuffer);
```

So in this way we can be more lenient on the usage of `ByteAccumulator` by supporting the deprecated way of using it, and now we have the replacement `ByteBufferAccumulator` which is more strict and must be used with an explicit call to `close()`.